src/freenet/clients/fcp/FCPPluginClient.java

```diff
@@ -0,0 +1,1189 @@
+     * of the message handling functions first as it puts additional constraints on the usage
+     * of the FCPPluginClient they receive.
+     * 
+     * @param direction
+     *            Whether to send the message to the server or the client message handler.<br><br>
+     * 
+     *            While you <b>can</b> use this to send messages to yourself, be careful not to
+     *            cause thread deadlocks with this. The function will call your message
+     *            handler function of {@link FredPluginFCPMessageHandler#handlePluginFCPMessage(
+     *            FCPPluginClient, FCPPluginMessage)} in <b>a different thread</b>, so it should not
+     *            cause deadlocks on its own, but you might produce deadlocks with your own thread
+     *            synchronization measures.<br><br>
+     * 
+     * @param message
+     *            You <b>must not</b> send the same message twice: This can break
+     *            {@link #sendSynchronous(SendDirection, FCPPluginMessage, long)}.<br>
+     *            To ensure this, always construct a fresh FCPPluginMessage object when re-sending
+     *            a message. If you use the constructor which allows specifying your own identifier,
+     *            always generate a fresh, random identifier.<br>
+     *            TODO: Code quality: Add a flag to FCPPluginMessage which marks the message as
+     *            sent and use it to log an error if someone tries to send the same message twice.
+     *            <br><br>
+     * 
+     * @throws IOException
+     *             If the connection has been closed meanwhile.<br/>
+     *             This FCPPluginClient <b>should be</b> considered as dead once this happens, you
+     *             should then discard it and obtain a fresh one.
+     * 
+     *             <p><b>ATTENTION:</b> If this is not thrown, that does NOT mean that the
+     *             connection is alive. Messages are sent asynchronously, so it can happen that a
+     *             closed connection is not detected before this function returns.<br/>
+     *             The only way of knowing that a send succeeded is by receiving a reply message
+     *             in your {@link FredPluginFCPMessageHandler}.<br>
+     *             If you need to know whether the send succeeded on the same thread which shall
+     *             call the send function, you can also use {@link #sendSynchronous(SendDirection,
+     *             FCPPluginMessage, long)} which will return the reply right away.</p>
+     * @see #sendSynchronous(SendDirection, FCPPluginMessage, long)
+     *          You may instead use the blocking sendSynchronous() if your thread needs to know
+     *          whether messages arrived, to ensure a certain order of arrival, or to know
+     *          the reply to a message.
+     */
+    public void send(final SendDirection direction, FCPPluginMessage message) throws IOException {
```

Comment:
Thhe requested improvements have been implemented as follows:

    Both ends are called the client? That there's some confusing naming!

FCPPluginClient has been renamed to FCPPluginConnection.
Also, it has been converted to from a class to an interface, which leaves room for changing the implementation in the future. This is also used for adding adapters as explained below....

    Should we really specify here in which direction we are about to send? It's rather obvious that the server only sends to the plugin, and the plugin only sends to the server, in all reasonable use cases. Maybe make this some kind of instance variable?

FCPPluginConnection now contains additional send() and sendSynchronous() which do not require a SendDirection.
They are implemented by wrapping the FCPPluginConnectionImpl in a "class DefaultSendDirectionAdapter implements FCPPluginConnection" which specifies the default direction. The wrapping is done automatically by the internal code, so plugin authors won't have to deal with it: The FCPPluginConnection objects they will be given will automatically be wrapped with the correct SendDirection depending on whether the FCPPluginConnection was handed out by API for servers or for clients. They won't even notice that they're using the adapter since the type which fred returns is the interface FCPPluginConnection, not the class of the adapter.